### PR TITLE
Release VM access hook

### PR DESCRIPTION
**Pull request by amicic**
Introduce a VM hook to trigger after VM access is released.
Allow user to enforce slow release VM access path (hence, enforce the
hook trigger) by setting J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS bit
(which is already used to enforce slow acquire VM access path, hence
renamed to a more generic name).

Signed-off-by: Aleksandar Micic <amicic@ca.ibm.com>

### CODE REVIEW

File: runtime/gc_glue_java/JNICriticalRegion.cpp [hunk 1]
```diff
@@ -120,6 +120,15 @@ MM_JNICriticalRegion::releaseAccess(J9VMThread* vmThread, UDATA* accessMask)
 		omrthread_monitor_exit(vm->exclusiveAccessMutex);
 	}
 
+	if (0 != (currentAccess & J9_PUBLIC_FLAGS_VM_ACCESS)) {
+		TRIGGER_J9HOOK_VM_RELEASEVMACCESS(vmThread->javaVM->hookInterface, vmThread);
+
+		/* Now that the hook has been invoked, allow inline VM access */
+		if(0 != (vmThread->publicFlags & J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS)) {
```
Comment (gacholio) [inline runtime/gc_glue_java/JNICriticalRegion.cpp]: blank after if

File: runtime/gc_glue_java/JNICriticalRegion.cpp [hunk 2]
```diff
@@ -55,9 +55,9 @@ MM_JNICriticalRegion::reacquireAccess(J9VMThread* vmThread, UDATA accessMask)
 	if (0 != (accessMask & J9_PUBLIC_FLAGS_VM_ACCESS)) {
 		TRIGGER_J9HOOK_VM_ACQUIREVMACCESS(vmThread->javaVM->hookInterface, vmThread);
 
-		/* Now that the hook has been invoked, allow inline VM access acquire */
-		if((vmThread->publicFlags & J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS_ACQUIRE) == J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS_ACQUIRE) {
-			clearEventFlag(vmThread, J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS_ACQUIRE);
+		/* Now that the hook has been invoked, allow inline VM access */
+		if(0 != (vmThread->publicFlags & J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS)) {
```
Comment (gacholio) [inline runtime/gc_glue_java/JNICriticalRegion.cpp]: blank after if

Comment (gacholio) [reply runtime/gc_glue_java/JNICriticalRegion.cpp]: sorry - gc code, feel free to ignore me

File: runtime/gc_glue_java/JNICriticalRegion.cpp (hunk 1 above)
Comment (gacholio) [reply runtime/gc_glue_java/JNICriticalRegion.cpp]: again, gc, but it's inconsistent with the if a few line above.

File: runtime/oti/j9vm.hdf [hunk 3]
```diff
@@ -267,6 +267,15 @@ typedef UDATA (* lookupNativeAddressCallback)(struct J9VMThread *currentThread,
 		<struct>J9VMAcquireVMAccessEvent</struct>
 		<data type="struct J9VMThread*" name="currentThread" description="current thread" />
 	</event>
+	
+	<event>
+		<name>J9HOOK_VM_RELEASEVMACCESS</name>
+		<description>
+				Triggered when a thread is about to release VM access.
```
Comment (gacholio) [inline runtime/oti/j9vm.hdf]: Probably worth documenting that the hook is called after access is released (which is the opposite of what the current comment implies).

File: runtime/vm/VMAccess.cpp [hunk 4]
```diff
@@ -385,9 +385,9 @@ void   internalAcquireVMAccessNoMutexWithMask(J9VMThread * vmThread, UDATA haltM
 
 	TRIGGER_J9HOOK_VM_ACQUIREVMACCESS(vm->hookInterface, vmThread);
 
-	/* Now that the hook has been invoked, allow inline VM access acquire */
-	if((vmThread->publicFlags & J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS_ACQUIRE) == J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS_ACQUIRE) {
-		VM_VMAccess::clearPublicFlags(vmThread, J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS_ACQUIRE);
+	/* Now that the hook has been invoked, allow inline VM access */
+	if(0 != (vmThread->publicFlags & J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS)) {
```
Comment (gacholio) [inline runtime/vm/VMAccess.cpp]: blank after if

File: runtime/vm/VMAccess.cpp [hunk 5]
```diff
@@ -456,6 +456,13 @@ static void  internalReleaseVMAccessNoMutexNoCheck(J9VMThread * vmThread)
 		omrthread_monitor_exit(vm->exclusiveAccessMutex);
 	}
 
+	TRIGGER_J9HOOK_VM_RELEASEVMACCESS(vm->hookInterface, vmThread);
+
+	/* Now that the hook has been invoked, allow inline VM access */
+	if(0 != (vmThread->publicFlags & J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS)) {
```
Comment (gacholio) [inline runtime/vm/VMAccess.cpp]: blank after if

File: runtime/vm/VMAccess.cpp [hunk 6]
```diff
@@ -632,9 +639,9 @@ IDATA   internalTryAcquireVMAccessNoMutexWithMask(J9VMThread * vmThread, UDATA h
 
 	TRIGGER_J9HOOK_VM_ACQUIREVMACCESS(vm->hookInterface, vmThread);
 
-	/* Now that the hook has been invoked, allow inline VM access acquire */
-	if((vmThread->publicFlags & J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS_ACQUIRE) == J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS_ACQUIRE) {
-		VM_VMAccess::clearPublicFlags(vmThread, J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS_ACQUIRE);
+	/* Now that the hook has been invoked, allow inline VM access */
+	if(0 != (vmThread->publicFlags & J9_PUBLIC_FLAGS_DISABLE_INLINE_VM_ACCESS)) {
```
Comment (gacholio) [inline runtime/vm/VMAccess.cpp]: blank after if

**Comment by gacholio** [comment]
jenkins test sanity xlinux jdk8

**Comment by gacholio** [comment]
@amicic Some conflicts here to be resolved.

**Comment by amicic** [comment]
@gacholio resolved

**Pull request merged**